pallets/polkapool/src/lib.rs

```diff
@@ -19,131 +19,160 @@
+use frame_support::pallet_prelude::*;
+use frame_support::sp_runtime::traits::AtLeast32BitUnsigned;
+use frame_support::traits::{Filter, Randomness};
 use frame_system::ensure_signed;
-use orml_traits::{MultiCurrency, MultiCurrencyExtended, MultiReservableCurrency};
+use orml_traits::{MultiCurrency, MultiReservableCurrency, MultiLockableCurrency};
 use polkadex_primitives::assets::AssetId;
 use polkadex_primitives::BlockNumber;
-use rand::{seq::SliceRandom, SeedableRng};
+use rand::{SeedableRng, seq::SliceRandom};
 use rand_chacha::ChaChaRng;
 use sp_arithmetic::traits::*;
 use sp_arithmetic::traits::{Bounded, One, SaturatedConversion, Saturating, Zero};
 use sp_core::H256;
 use sp_std::boxed::Box;
+use sp_std::collections::vec_deque::VecDeque;
 use sp_std::vec::Vec;
+use sp_std::vec;
 
 /// Configure the pallet by specifying the parameters and types on which it depends.
 pub trait Config: frame_system::Config {
     /// Because this pallet emits events, it depends on the runtime's definition of an event.
     type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
     /// The aggregated origin which the dispatch will take.
-    type Origin: OriginTrait<PalletsOrigin = Self::PalletsOrigin>
-        + From<Self::PalletsOrigin>
-        + IsType<<Self as frame_system::Config>::Origin>;
+    type Origin: OriginTrait<PalletsOrigin=Self::PalletsOrigin>
+    + From<Self::PalletsOrigin>
+    + IsType<<Self as frame_system::Config>::Origin>;
 
     /// The caller origin, overarching type of all pallets origins.
     type PalletsOrigin: From<frame_system::RawOrigin<Self::AccountId>> + Codec + Clone + Eq;
 
     /// The aggregated call type.
     type Call: Parameter
-        + Dispatchable<Origin = <Self as Config>::Origin>
-        + GetDispatchInfo
-        + From<frame_system::Call<Self>>;
+    + Dispatchable<Origin=<Self as Config>::Origin>
+    + GetDispatchInfo
+    + From<frame_system::Call<Self>>;
     /// Balance Type
     type Balance: Parameter
-        + Member
-        + AtLeast32BitUnsigned
-        + Default
-        + Copy
-        + MaybeSerializeDeserialize
-        + Clone
-        + Zero
-        + One
-        + PartialOrd
-        + Bounded;
+    + Member
+    + AtLeast32BitUnsigned
+    + Default
+    + Copy
+    + MaybeSerializeDeserialize
+    + Clone
+    + Zero
+    + One
+    + PartialOrd
+    + Bounded;
     /// Module that handles tokens
     type Currency: MultiReservableCurrency<
         Self::AccountId,
-        CurrencyId = AssetId,
-        Balance = Self::Balance,
-    >;
+        CurrencyId=AssetId,
+        Balance=Self::Balance,
+    > + MultiLockableCurrency<Self::AccountId>;
     /// Min amount that must be staked
     type MinStakeAmount: Get<Self::Balance>;
     /// Maximum allowed Feeless Transactions in a block
     type MaxAllowedWeight: Get<Weight>;
     /// Min Stake Period
-    type MinStakePeriodPerWeight: Get<u32>;
+    type MinStakePeriod: Get<BlockNumber>;
```

Comment:
Defines the min amount of time the user needs to stake their tokens, currently set to 28 days